src/vm/method.hpp

```diff
@@ -1634,11 +1634,13 @@ class MethodDesc
     //     pDispatchingMT - method table of the object that the method is being dispatched on, can be NULL.
     //     fFullBackPatch - indicates whether to patch all possible slots, including the ones 
     //                      expensive to patch
+    //     pPreviousEntryPoint - if this is not NULL, the value is used as the expected value for all the patching
```

Comment:
Is `pPreviousEntryPoint` just an optimization, or is it required for correctness?

I am worried that there can be bad (crashing?) races between `Prestub` and `DoBackpatch` because of this.

Reply:
My understanding is that this is actually required for correctness.
Basically DoBackpatch needs to determine the expected "old" value of the method entry point. For method with temporary entry point we use that - assuming that the entry point only ever changes from temporary to stable and never again. So we blindly take the temporary and see if something matches it - if so we patch that.
With the zapped methods, we don't have a place to lookup the address of the zapped precode, since by the time we get to DoBackpatch the method slot is already overwritten with the stable entry point. So I had to work around that.
The way it works is that SetStableEntryPointInterlocked returns the previous value (the zapped precode, we actually verify this in the method itself). This previous value is then passed to DoBackpatch. DoBackpatch will then use it as the temporary entry point, that is it will see if there are other slots with the previous value, and if so, it will patch them with the stable entry point.
This actually works for both zapped and non-zapped methods, since the logic in SetStableEntryPointInterlocked must be the same as the one in DoBackpatch.

Regarding races:
- SetStableEntryPointInterlocked will return temporary entry point for non-zapped method in this out-parameter. So the behavior is exactly the same as before (we just don't load it again in DoBackpatch, but take it as a parameter instead).
- SetStableEntryPointInterlocked will return the previous slot value for zapped methods, but only if it passes the IsZappedPrecode condition. So if there's a race and the stable entry point is set twice, the second call will see the new value in the slot, but that will not pass the IsZappedPrecode - in which case we would return NULL as the previous value -> leave DoBackpatch do what it did so far (which for zapped methods is effectively nothing in this case since the method is not marked as having a temporary entry point).
- If it so happens that the second call to SetStableEntryPointInterlocked still sees the old value there - the zapped precode - it will return that. DoBackpatch will use it just like it does temporary entry point, potentially trying to backpatch the method for the second time. But since it does so by reading the stable entry point and then patching with that, at worst it would patch to the same value - in reality, that is unlikely to happen, since the old value will not match what's in the slots anymore.